[SYCL][Doc] Update compiler options descriptions (#3340)
intel/llvm: sycl/doc/UsersManual.md

    Tells the compiler to generate code for specified devices. T is a target
    triple device name. You can specify more than one target, comma separated.
    The following triples are supported:
    * spir64-unknown-unknown-sycldevice - this is the default generic SPIR-V
      target;
    * spir64_x86_64-unknown-unknown-sycldevice - generate code ahead of time
      for x86_64 CPUs;
    * spir64_fpga-unknown-unknown-sycldevice - generate code ahead of time for
      Intel FPGA;
    * spir64_gen-unknown-unknown-sycldevice - generate code ahead of time for
      Intel Processor Graphics;
    * nvptx64-nvidia-cuda-sycldevice - generate code ahead of time for CUDA
      target;

## Language options

**`-sycl-std=<value>`** [EXPERIMENTAL]

    SYCL language standard to compile for. Possible values:
    * 121 - SYCL 1.2.1
    * 2020 - SYCL 2020
    It doesn't guarantee specific standard compliance, but some selected
    compiler features change behavior.
    It is under development and not recommended to use in production
    environment.
    Default value is 2020.

**`-f[no-]sycl-unnamed-lambda`**

    Enables/Disables unnamed SYCL lambda kernels support.

>     Enabled by default.

Reviewer: I don't think it's enabled by default. @mdtoguchi, am I right?

Author: I was thinking it's on under sycl-std=2020, but I could be wrong. @mdtoguchi?

Reviewer: This is not what I see in the source code: https://github.com/intel/llvm/blob/sycl/clang/lib/Driver/ToolChains/Clang.cpp#L6589.

Reviewer 2: Right, this is not enabled by default.

Author: Thanks. Fixed.

# Users Manual

This is the list of SYCL specific options supported by compiler and some
examples.

Options marked as [DEPRECATED] are going to be removed in some future updates.
Options marked as [EXPERIMENTAL] are expected to be used only in limitted cases
and not recommended to use in production environment.

## Generic options

**`-fsycl`**

    General enabling option for SYCL compilation and linking mode. List of
    targets can be specified with `-fsycl-targets`. This is fundamental option
    for any SYCL compilation. All other SYCL specific options require it.

**`-fsycl-targets=<T1>[,...,<Tn>]`**

>     Tells the compiler to generate code for specified devices. T is a target
>     triple device name. You can specify more than one target, comma separated.

Reviewer 3: I think it is important to associate the widely used "ahead of time compilation" term with this option. We mention it a lot in various other docs.
```suggestion
    Enables "ahead of time" compilation ("AOT") for given device targets.
    Tells the compiler to generate code for specified devices. `T...` is an LLVM target
    triple-based string representing a target device architecture. You can specify more than one target, comma separated.
```

Author: Rephrased to focus on enabling part.

      Intel Processor Graphics;
    Available in special build configuration:
    * nvptx64-nvidia-cuda-sycldevice - generate code ahead of time for CUDA
      target;

## Language options

**`-sycl-std=<value>`** [EXPERIMENTAL]

    SYCL language standard to compile for. Possible values:
    * 121 - SYCL 1.2.1
    * 2020 - SYCL 2020
    It doesn't guarantee specific standard compliance, but some selected
    compiler features change behavior.
    It is under development and not recommended to use in production
    environment.
    Default value is 2020.

**`-f[no-]sycl-unnamed-lambda`**

    Enables/Disables unnamed SYCL lambda kernels support.
    Disabled by default.

## Optimization options

**`-f[no-]sycl-early-optimizations`**

    Enables (or disables) intermediate representation optimization pipeline
    before translation to SPIR-V. Have effect only if optimizations are turned
    on by standard compiler options (-O1 or higher).
    Enabled by default.

**`-f[no-]sycl-dead-args-optimization`**

    Enables (or disables) LLVM IR dead argument elimination pass to remove
    unused arguments for the kernel functions before translation to SPIR-V.

>     Disabled by default.

Reviewer 3: some targets (PTX) don't use SPIRV, need to specify option's effect for them as well. @bader, do you know if this has effect for PTX?

Reviewer: IIRC, this optimization impacts only SPIR-targeted code. Relevant issue - https://github.com/intel/llvm/issues/2359.

Reviewer 3: Thanks. Then this could be rephrased as "... remove unused arguments for the kernel functions before translation to SPIR-V.
Has no effect on non-SPIRV targets. Disabled by default."

Author: Added.

**`-f[no-]sycl-id-queries-fit-in-int`**

    Assume/Do not assume that SYCL ID queries fit within MAX_INT. It assumes
    that these values fit within MAX_INT:
    * id class get() member function and operator[]
    * item class get_id() member function and operator[]
    * nd_item class get_global_id()/get_global_linear_id() member functions
    Enabled by default.

## Target toolchain options

**`-Xsycl-target-backend=<T> "options"`**
**`-Xs "options"`**

    Pass "options" to the backend of target device compiler, specified by
    triple T. The backend of device compiler generates target machine code from
    intermediate representation. This option can be used to tune code

>     generation for a specific target. The "options" are used during offline
>     compilation and are also saved in a fat binary for online compilation.

Reviewer 3: AFAIR, options are only passed through the fat binary for the generic SPIRV target.
+ @mdtoguchi to confirm.
```suggestion
    generation for a specific target. The "options" are used during offline
    compilation and are also saved in a fat binary for online compilation (for the `spir64-unknown-unknown-sycldevice` generic target).
```

Reviewer 2: Yes, that is correct - only for the generic default target.

Author: Rephrased.

    '-fsycl-targets' option (or default one).

**`-Xsycl-target-frontend=<T> "options"`**

    Pass "options" to the frontend of target device compiler, specified by
    triple T. This option can be used to control of intermediate representation
    generation during offline or online compilation.

**`-Xsycl-target-linker=<T> "options"`**

    Pass "options" to the device code linker, when linking multiple device
    object modules. T is specific target device triple.

## Link options

**`-fsycl-link`**

    Link device object modules and wrap those into a host-compatible object
    module that can be linked later by any standard host linker into the final
    fat binary.

**`-fsycl-link-targets=<T1,...,Tn>`** [DEPRECATED]

    Specify comma-separated list of triples SYCL offloading targets to produce
    linked device images. Used in a link step to link device code for given
    targets and output multiple linked device code images, whose names consist
    of the common prefix taken from the -o option and the triple string.
    Does not produce fat binary and must be used together with -fsycl.

**`-fsycl-add-targets=<T1:file1...Tn:filen>`** [DEPRECATED]

> 

Reviewer 2: For `-fsycl-[link|add]-targets` are there any current users of these?  It would seem the desired approach is to use our regular AOT compilation model.

Author: Confirmed with FPGA folks, should be OK to deprecate.

**`-f[no-]sycl-early-optimizations`**

    Enables (or disables) intermediate representation optimization pipeline
    before translation to SPIR-V. Have effect only if optimizations are turned
    on by standard compiler options (-O1 or higher).
    Enabled by default.

**`-f[no-]sycl-dead-args-optimization`**

    Enables (or disables) LLVM IR dead argument elimination pass to remove
    unused arguments for the kernel functions before translation to SPIR-V.
    Disabled by default.

**`-f[no-]sycl-id-queries-fit-in-int`**

    Assume/Do not assume that SYCL ID queries fit within MAX_INT. It assumes
    that these values fit within MAX_INT:
    * id class get() member function and operator[]
    * item class get_id() member function and operator[]
    * nd_item class get_global_id()/get_global_linear_id() member functions
    Enabled by default.

## Target toolchain options

**`-Xsycl-target-backend=<T> "options"`**
**`-Xs "options"`**

    Pass "options" to the backend of target device compiler, specified by
    triple T. The backend of device compiler generates target machine code from
    intermediate representation. This option can be used to tune code

>     generation for a specific target. For offline compilation it happens during
>     compilation. For online compilation options are saved in a fat binary to be
>     used later, during runtime compilation.

Reviewer 2: ```suggestion
    generation for a specific target. The options are used during offline compilation
    and are also saved in a fat binary when used for online compilation to be used
    at runtime compilation.
```

>     -Xs is a shortcut to pass options to all backends specified (or default

Reviewer 2: ```suggestion
    -Xs is a shortcut to pass "options" to all backends specified via the
    '-fsycl-targets' option (or default
```

    FPGA and turns on the debug options that are needed to generate FPGA
    reports. It is functionally equivalent shortcut to
    `-fsycl-targets=spir64_fpga-unknown-unknown-sycldevice -g -MMD` on Linux
    and `-fsycl-targets=spir64_fpga-unknown-unknown-sycldevice -Zi -MMD` on
    Windows.

**`-fsycl-link=<output>`**

    Controls FPGA target binary output format. Same as -fsycl-link, but
    optional output can be one of the following:
    * early - generate html reports and an intermediate object file that avoids
    a full Quartus compile. Usually takes minutes to generate. Link can later
    be resumed from this point using -fsycl-link=image.
    * image - generate a bitstream which is ready to be linked and used on a
    FPGA board. Usually takes hours to generate.

**`-reuse-exe=<exe>`**

    Speed up FPGA backend compilation if the device code in <binary> is
    unchanged. If it's safe to do so the compiler will re-use the device binary
    embedded within it. This can be used to minimize or avoid long Quartus
    compile times for FPGA targets when the device code is unchanged.

## Other options

**`-fsycl-device-only`**

>     Compile only device part of the code and discard host part.

Reviewer: Probably it would be more clear to say that host compilation is skipped in this mode.

Author: Replaced to "ignore".

> **`-fsycl-use-bitcode`** [DEPRECATED]

Reviewer: I found this option quite useful for development/debug purposes. What is the suggested replacement? `-emit-llvm`?

Author: @mdtoguchi, is it a good replacement?

Reviewer 2: `-fsycl-use-bitcode` is the default, so there is only `-fno-sycl-use-bitcode` that is useful when paired with `-fsycl-device-only`.  Is it possible for this to not be deprecated and stick around?  There is desire to provide support for .spv and .bc all the way to fat obj.

Author: Ok. Updated.

    `-fsycl-targets=spir64_fpga-unknown-unknown-sycldevice -g -MMD` on Linux
    and `-fsycl-targets=spir64_fpga-unknown-unknown-sycldevice -Zi -MMD` on
    Windows.

**`-fsycl-link=<output>`**

    Controls FPGA target binary output format. Same as -fsycl-link, but
    optional output can be one of the following:
    * early - generate html reports and an intermediate object file that avoids
    a full Quartus compile. Usually takes minutes to generate. Link can later
    be resumed from this point using -fsycl-link=image.
    * image - generate a bitstream which is ready to be linked and used on a
    FPGA board. Usually takes hours to generate.

**`-reuse-exe=<exe>`**

    Speed up FPGA backend compilation if the device code in <binary> is
    unchanged. If it's safe to do so the compiler will re-use the device binary
    embedded within it. This can be used to minimize or avoid long Quartus
    compile times for FPGA targets when the device code is unchanged.

## Other options

**`-fsycl-device-only`**

    Compile only device part of the code and ignore host part.

**`-f[no-]sycl-use-bitcode`**

    Emit SYCL device code in LLVM-IR bitcode format. When disabled, SPIR-V is

>     emitted.

Reviewer 3: We don't test `-fno-sycl-use-bitcode` so this is 'use at your own risk' option. Will likely cause problems in scenarios with real device code linkage.

Author: Marked it [EXPERIMENTAL]
